syntax = "proto3";

package mathgame.protobuf;

message Response {
    enum Type {
        NEW_PLAYER_JOINED = 0;
        GAME_STARTED = 1; // ゲーム開始
        PROBLEM = 2; // 問題提示
        ANSWER_RESULT = 3; // 答えた結果
        PROBLEM_RESULT = 4; // 問題の結果
        GAME_RESULT = 5; // ゲーム終了
    }
    Type type = 1;
    bool broadcast = 2;
    string destination_id = 3;

    oneof payload {
        NewPlayerJoined new_player_joined = 10;
        Problem problem = 11;
        AnswerResult answer_result = 12;
        ProblemResult problem_result = 13;
        GameResult game_result = 14;
    }
}

message NewPlayerJoined {
    string player_id = 1;
}

message Problem {
    enum Type {
        ADDITION = 0;
    }
    Type type = 1;
    oneof payload {
        Addition addition = 10;
    }
}

message Addition {
    int32 x = 1;
    int32 y = 2;
}

message AnswerResult {
    bool correct = 1;
    int32 score = 2;
}

message ProblemResult {
    repeated PlayerScore player_scores = 1;
}

message PlayerScore {
    string player_id = 1;
    string score = 2;
}

message GameResult {
    string winner = 1;
}
